cmd/consumer/main.go

```diff
@@ -0,0 +1,175 @@
+package main
+
+import (
+	"encoding/base64"
+	"fmt"
+	"log"
+	"math/big"
+	"os"
+	"time"
+
+	"github.com/Clever/amazon-kinesis-client-go/kcl"
+
+	"github.com/Clever/kinesis-to-firehose/firehose"
+)
+
+type RecordProcessor struct {
+	shardID           string
+	sleepDuration     time.Duration
+	checkpointRetries int
+	checkpointFreq    time.Duration
+	largestSeq        *big.Int
+	largestSubSeq     int
+	lastCheckpoint    time.Time
+	firehoseWriter    *firehose.FirehoseWriter
+}
+
+func New() *RecordProcessor {
+	return &RecordProcessor{
+		sleepDuration:     5 * time.Second,
+		checkpointRetries: 5,
+		checkpointFreq:    60 * time.Second,
+	}
+}
+
+func (rp *RecordProcessor) Initialize(shardID string) error {
+	rp.shardID = shardID
+	rp.lastCheckpoint = time.Now()
+	return nil
+}
+
+func (rp *RecordProcessor) checkpoint(checkpointer kcl.Checkpointer, sequenceNumber string, subSequenceNumber int) {
+	for n := 0; n < rp.checkpointRetries; n++ {
+		if err := checkpointer.Checkpoint(sequenceNumber, subSequenceNumber); err == nil {
+			break
+		} else {
+			if cperr, ok := err.(kcl.CheckpointError); ok {
+				switch cperr.Error() {
+				case "ShutdownException":
+					fmt.Fprintf(os.Stderr, "Encountered shutdown exception, skipping checkpoint\n")
+					return
+				case "ThrottlingException":
+					if rp.checkpointRetries-1 == n {
+						fmt.Fprintf(os.Stderr, "Failed to checkpoint after %d attempts, giving up.\n", n)
+						return
+					}
+					fmt.Fprintf(os.Stderr, "Was throttled while checkpointing, will attempt again in %s", rp.sleepDuration)
+				case "InvalidStateException":
+					fmt.Fprintf(os.Stderr, "MultiLangDaemon reported an invalid state while checkpointing\n")
+				}
+			}
+			fmt.Fprintf(os.Stderr, "Encountered an error while checkpointing: %s", err)
+		}
+		time.Sleep(rp.sleepDuration)
+	}
+}
+
+func (rp *RecordProcessor) shouldUpdateSequence(sequenceNumber *big.Int, subSequenceNumber int) bool {
+	return rp.largestSeq == nil || sequenceNumber.Cmp(rp.largestSeq) == 1 ||
+		(sequenceNumber.Cmp(rp.largestSeq) == 0 && subSequenceNumber > rp.largestSubSeq)
+}
+
+func (rp *RecordProcessor) ProcessRecords(records []kcl.Record, checkpointer kcl.Checkpointer) error {
+	for _, record := range records {
+		// Base64 decode the record
+		data, err := base64.StdEncoding.DecodeString(record.Data)
+		if err != nil {
+			return err
+		}
+		msg := string(data)
+
+		// TODO: Add additional "decoding"
+		// - pull JSON data out of string
+		// - parse RSyslog format
+		// - etc...
+
+		// Write the message to firehose
+		err = rp.firehoseWriter.ProcessMessage(msg)
+		if err != nil {
+			return err
+		}
+
+		// Handle checkpointing
+		// TODO: How to handle the difference between ProcessMessage (sent to FirehoseOutput) vs successfully sent?
```

Comment:
My main concern here is if the app crashes.

As soon as a message is handed off to the `batcher`, it's considered processed. However, if they batcher holds onto it (say for 1 minute), it's possible that the app could crash during this time. If a checkpoint also occurs, KCL will consider those messages processed, even though our `batcher` never even attempted to send them.

One idea I considered is whether we might pursue something like "deliver" (handed to batcher) vs "ack" (batcher confirms messages was processed successfully)... here inspired by [RabbitMQ confirmations](https://www.rabbitmq.com/confirms.html)